src/main/java/com/redhat/devtools/intellij/common/utils/NetworkUtils.java

```diff
@@ -0,0 +1,147 @@
+import com.intellij.util.net.HttpConfigurable;
+import com.intellij.util.net.IdeaWideAuthenticator;
+import com.intellij.util.net.IdeaWideProxySelector;
+import okhttp3.Authenticator;
+import okhttp3.OkHttpClient;
+import org.jetbrains.annotations.NotNull;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.PasswordAuthentication;
+import java.net.Proxy;
+import java.net.SocketAddress;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.intellij.openapi.util.text.StringUtil.isNotEmpty;
+import static okhttp3.Credentials.basic;
+
+public class NetworkUtils {
+
+    public static OkHttpClient getClient() {
+        final HttpConfigurable httpConfigurable = HttpConfigurable.getInstance();
+        final IdeaWideProxySelector ideaWideProxySelector = new IdeaWideProxySelector(httpConfigurable);
+        final IdeaWideAuthenticator ideaWideAuthenticator = new IdeaWideAuthenticator(httpConfigurable);
+        final Authenticator proxyAuthenticator = getProxyAuthenticator(ideaWideAuthenticator);
+        final OkHttpClient.Builder builder = new OkHttpClient.Builder();
+
+        builder.proxySelector(ideaWideProxySelector)
+                .proxyAuthenticator(proxyAuthenticator);
+
+        return builder.build();
+    }
+
+    private static Authenticator getProxyAuthenticator(IdeaWideAuthenticator ideaWideAuthenticator) {
+        Authenticator proxyAuthenticator = null;
+
+        if (Objects.nonNull(ideaWideAuthenticator)) {
+            proxyAuthenticator = (route, response) -> {
+                final PasswordAuthentication authentication = ideaWideAuthenticator.getPasswordAuthentication();
+                final String credential = basic(authentication.getUserName(), Arrays.toString(authentication.getPassword()));
+                return response.request().newBuilder()
+                        .header("Proxy-Authorization", credential)
+                        .build();
+            };
+        }
+
+        return proxyAuthenticator;
+    }
+
+    @NotNull
+    public static Map<String, String> buildEnvironmentVariables(String url) throws URISyntaxException {
+        final int FIRST = 0;
+        final String HTTP_PROXY = "HTTP_PROXY";
+        final String HTTPS_PROXY = "HTTPS_PROXY";
+        final String ALL_PROXY = "ALL_PROXY";
+        final Set<String> proxyEnvironmentVariables = Sets.newHashSet(HTTP_PROXY, HTTPS_PROXY, ALL_PROXY);
+
+        final Map<String, String> environmentVariables = new HashMap<>(6);
+
+        final HttpConfigurable httpConfigurable = HttpConfigurable.getInstance();
+        final IdeaWideProxySelector ideaWideProxySelector = new IdeaWideProxySelector(httpConfigurable);
+        final URI uri = new URI(url);
+        final List<Proxy> proxies = ideaWideProxySelector.select(uri);
```

Comment:
I don't get how this works. The HttpConfigurable is initialized with the default values for all its flags (e.g the USE_HTTP_PROXY set to false) and these flags are used inside the `select` method to create the proxies list. But if the flags values are never updated the output will always be the same. Am i missing something?

Reply:
the HttpConfigurable represent the IJ HTTP settings as configured by the user. So if user update those settings you will get a list of proxies as a result of select